<center>
<img src="https://drive.google.com/uc?id=1f1gGVI-rxcHjA90WEGNvvtSXF1pAxQwg" alt="Fasilkom UI" width="300"/>

CSGE603130 • Kecerdasan Artifisial dan Sains Data Dasar

Semester Gasal 2023/2024

Fakultas Ilmu Komputer, Universitas Indonesia

##**Lab 1: *Introduction to Data Science Tools***

###**Tenggat Waktu: 9 September 2023, 23.55 WIB**
</center>

### **Ketentuan:**

1. Dokumen template lab dengan format .ipynb dan dataset (jika dibutuhkan) telah disediakan di SCeLe
2. Jalankan kode pada dokumen .ipynb dan perhatikan dengan saksama apa yang potongan kode tersebut lakukan beserta dengan keluarannya. Jawablah **pertanyaan** yang disisipkan pada potongan program yang diberikan.
3. Dokumen Jupyter Notebook yang telah dilengkapi dengan jawaban dikumpulkan dengan format penamaan **Kelas_LabX_NPM_Nama.ipynb** . Contoh: A_Lab1_1806205773_Abdurrafi Arief.ipynb
4. Kumpulkan dokumen tersebut pada submisi yang telah disediakan di SCeLe sesuai dengan kelas masing-masing sebelum **Tenggat Waktu: 9 September 2023, 23.55 WIB**. Keterlambatan pengumpulan akan dikenakan pinalti.
5. Lab ini dirancang sebagai **tugas mandiri**. Plagiarisme tidak diperkenankan dalam bentuk apapun. Adapun kolaborasi berupa diskusi (tanpa menyalin maupun mengambil jawaban orang lain) dan literasi masih diperbolehkan dengan mencantumkan kolaborator dan sumber.



# A. Pengenalan Pandas
[Pandas](https://pandas.pydata.org/) adalah library Python yang digunakan untuk analisis data. Pandas menyediakan struktur data dan fungsi high-level untuk membuat pekerjaan dengan data terstruktur/tabular lebih mudah. Struktur data utama yang digunakan di Pandas adalah **Series** dan **DataFrame**. **Series adalah array 1 dimensi** yang berisi data dan index. **DataFrame adalah array 2 dimensi** yang berisi data dan index untuk baris dan kolom.


Dengan menggunakan library Pandas, kita dapat melakukan berbagai operasi seperti:
- Memuat data dari berbagai format file seperti CSV, Excel, dan SQL
- Memilih kolom dan baris tertentu dari data
- Menggabungkan data dari beberapa sumber
- Menghapus dan mengisi nilai yang hilang
- Melakukan perhitungan statistik deskriptif seperti mean, median, dan standard deviation


Secara umum, Pandas adalah library yang berguna bagi siapa pun yang bekerja dengan data dan membutuhkan alat untuk memanipulasi dan menganalisis data.

In [ ]:
# import library pandas
import pandas as pd

## A.1. Series
**Series** adalah data 1 dimensi yang mirip dengan array. Series memiliki index dan value. Index secara default adalah angka dari 0 sampai N-1, dimana N adalah panjang Series. Series dapat dibuat dari list, tuple, dan dictionary.

Salah satu cara yang umum digunakan untuk membuat objek **Series** adalah sebagai berikut.
```
pd.Series(data)
```
`data` dapat berupa list, tuple, atau dictionary.

In [ ]:
# Membuat objek Series dari sebuah list
series = pd.Series([1,4,9,16,25,36])
series

0     1
1     4
2     9
3    16
4    25
5    36
dtype: int64

Seperti yang dapat dilihat di atas, Series memiliki index (sebelah kiri) dan value (sebelah kanan). Kedua hal tersebut dapat diakses dengan menggunakan atribut `index` dan `values`.

In [ ]:
series.index

RangeIndex(start=0, stop=6, step=1)

In [ ]:
series.values

array([ 1,  4,  9, 16, 25, 36], dtype=int64)

Sama seperti list pada Python, Series pada Pandas dapat diakses menggunakan indexing dan slicing.

In [ ]:
series[3]

16

In [ ]:
series[2:4]

2     9
3    16
dtype: int64

## A.2. DataFrame
**DataFrame** adalah data 2 dimensi yang mirip dengan tabel. DataFrame memiliki index dan column. Index secara default adalah angka dari 0 sampai N-1, dimana N adalah panjang DataFrame. DataFrame dapat dibuat dari list, tuple, dictionary, dan Series. Setiap kolom pada dataframe dapat memiliki tipe data yang berbeda, seperti integer, string, atau float. Mirip dengan Series, Dataframe juga memiliki index unik yang digunakan untuk mengidentifikasi setiap barisnya.

**DataFrame** dapat digunakan untuk menyimpan dan memanipulasi data yang dibuat dari berbagai sumber, seperti CSV, Excel, dan SQL. Salah satu cara yang umum digunakan untuk membuat objek **DataFrame** adalah sebagai berikut.
```
pd.DataFrame(data, columns=column_names)
```
`data` dapat berupa list, tuple, dictionary, dan Series.

`column_names` adalah list yang berisi nama kolom.

Berikut adalah beberapa cara untuk membuat DataFrame.

### A.2.1. Membuat DataFrame dari Series

In [ ]:
series = pd.Series(["apel", "mangga", "jeruk", "lemon"])
series

In [ ]:
df = pd.DataFrame(series, columns=["Buah"])
df

Buah
0    apel
1  mangga
2   jeruk
3   lemon

### A.2.2. Membuat DataFrame dari List of Dictionary

In [ ]:
data = [
    {"nama": "Didi", "umur": 21},
    {"nama": "Dede", "umur": 22},
    {"nama": "Dodo", "umur": 23}
]
df = pd.DataFrame(data)
df

nama  umur
0  Didi    21
1  Dede    22
2  Dodo    23

### A.2.3. Membuat DataFrame dari Dictionary of Series

In [ ]:
indonesia = pd.Series(["apel", "mangga", "jeruk", "lemon"])
indonesia

0      apel
1    mangga
2     jeruk
3     lemon
dtype: object

In [ ]:
english = pd.Series(["apple", "mango", "orange", "lemon"])
english

0     apple
1     mango
2    orange
3     lemon
dtype: object

In [ ]:
df = pd.DataFrame({
    "indonesia": indonesia,
    "english": english
})
df

indonesia english
0      apel   apple
1    mangga   mango
2     jeruk  orange
3     lemon   lemon

### A.2.4. Membuat DataFrame dari Sebuah File

#### Menggunakan Path Remote (URL)

In [ ]:
# Membuat DataFrame dari file di internet
df_remote = pd.read_csv(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/glass/glass.data", # URL file
    header=None, # tidak ada baris header
)
df_remote

0        1      2     3     4      5     6     7     8    9   10
0      1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0   1
1      2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0   1
2      3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0   1
3      4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0   1
4      5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0   1
..   ...      ...    ...   ...   ...    ...   ...   ...   ...  ...  ..
209  210  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0   7
210  211  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0   7
211  212  1.52065  14.36  0.00  2.02  73.42  0.00  8.44  1.64  0.0   7
212  213  1.51651  14.38  0.00  1.94  73.61  0.00  8.48  1.57  0.0   7
213  214  1.51711  14.23  0.00  2.08  73.36  0.00  8.62  1.67  0.0   7

[214 rows x 11 columns]

**Memberikan nama kolom pada DataFrame**

In [ ]:
df_remote.columns = [
    'Id number', 'RI', 'Na', 'Mg',
    'Al', 'Si', 'K', 'Ca',
    'Ba', 'Fe', 'Type-of-glass'
]
df_remote

Id number       RI     Na    Mg    Al     Si     K    Ca    Ba   Fe  \
0            1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0   
1            2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0   
2            3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0   
3            4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0   
4            5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0   
..         ...      ...    ...   ...   ...    ...   ...   ...   ...  ...   
209        210  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0   
210        211  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0   
211        212  1.52065  14.36  0.00  2.02  73.42  0.00  8.44  1.64  0.0   
212        213  1.51651  14.38  0.00  1.94  73.61  0.00  8.48  1.57  0.0   
213        214  1.51711  14.23  0.00  2.08  73.36  0.00  8.62  1.67  0.0   

     Type-of-glass  
0                1  
1                1  
2                1  
3                1  
4                1  
..             ...  
209              7  
210              7  
211              7  
212              7  
213              7  

[214 rows x 11 columns]

**Deskripsi Dataset "glass"**
- Id number: 1 to 214
- RI: refractive index
- Na: Sodium (unit measurement: weight percent in corresponding oxide, as are attributes 4-10)
- Mg: Magnesium
- Al: Aluminum
- Si: Silicon
- K: Potassium
- Ca: Calcium
- Ba: Barium
- Fe: Iron
- Type-of-glass: (class attribute)


Sumber data set: https://archive.ics.uci.edu/ml/datasets/Glass+Identification

#### Menggunakan Path Lokal

In [ ]:
df_local = pd.read_csv(
    "./iris.csv" # relative path dari file csv di lokal
    # tidak menggunakan header=None karena file csv sudah menyertakan header
)
df_local

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

### A.2.5. Melihat Cuplikan DataFrame

Menggunakan fungsi `head(n)` untuk melihat 5 baris pertama dari DataFrame. Parameter `n` dapat digunakan untuk menentukan jumlah baris yang ingin ditampilkan. Jika tidak diberikan nilai, maka nilai defaultnya adalah 5. Untuk melihat 5 baris terakhir dari DataFrame, dapat menggunakan fungsi `tail(n)`.

In [ ]:
df = df_local

# Menampilkan 5 baris pertama dari DataFrame
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [ ]:
# Menampilkan 3 baris pertama dari DataFrame
df.head(3)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa

In [ ]:
# menampilkan 5 baris terakhir dari DataFrame
df.tail()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica

In [ ]:
# Menampilkan 5 baris random dari DataFrame
df.sample(5)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
138  139            6.0           3.0            4.8           1.8   
112  113            6.8           3.0            5.5           2.1   
51    52            6.4           3.2            4.5           1.5   
80    81            5.5           2.4            3.8           1.1   
83    84            6.0           2.7            5.1           1.6   

             Species  
138   Iris-virginica  
112   Iris-virginica  
51   Iris-versicolor  
80   Iris-versicolor  
83   Iris-versicolor

### A.2.6. Melihat Informasi DataFrame

Menampilkan dimensi dataframe menggunakan atribut `shape`.

In [ ]:
print(f"Dimensi dari dataframe iris adalah {df.shape}")
print(f"Terdiri dari {df.shape[0]} baris dan {df.shape[1]} kolom")

Dimensi dari dataframe iris adalah (150, 6)
Terdiri dari 150 baris dan 6 kolom


Menampilkan tipe data untuk setiap kolom menggunakan atribut `dtypes`.

In [ ]:
df.dtypes

Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object

Menampilkan statistik deskriptif untuk setiap kolom menggunakan fungsi `describe()`.

In [ ]:
df.describe()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000

In [ ]:
# untuk kolom tertentu
df['SepalLengthCm'].describe()

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: SepalLengthCm, dtype: float64

Menghitung jumlah nilai yang tidak ada (missing values) untuk setiap kolom menggunakan fungsi `isnull().sum()`.

In [ ]:
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

### A.2.7. Mengakses Kolom dan Baris pada DataFrame

#### Mengakses kolom menggunakan square bracket
Kita dapat mengakses kolom DataFrame seperti mengakses indeks pada array atau list, yaitu dengan Square Bracket, baik sebagai Series maupun DataFrame baru.

In [ ]:
# Mengakses kolom sebagai Series
col_1 = df['PetalLengthCm']
print(f"tipe objek: {type(col_1)}")
col_1

tipe objek: <class 'pandas.core.series.Series'>


0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: PetalLengthCm, Length: 150, dtype: float64

In [ ]:
# Mengakses kolom sebagai DataFrame
col_1 = df[['PetalLengthCm']]
print(f"tipe objek: {type(col_1)}")
col_1

tipe objek: <class 'pandas.core.frame.DataFrame'>


PetalLengthCm
0              1.4
1              1.4
2              1.3
3              1.5
4              1.4
..             ...
145            5.2
146            5.0
147            5.2
148            5.4
149            5.1

[150 rows x 1 columns]

In [ ]:
# Mengakses lebih dari 1 kolom sebagai DataFrame
col_1_2 = df[['PetalLengthCm', 'PetalWidthCm']]
print(f"tipe objek: {type(col_1_2)}")
col_1_2

tipe objek: <class 'pandas.core.frame.DataFrame'>


PetalLengthCm  PetalWidthCm
0              1.4           0.2
1              1.4           0.2
2              1.3           0.2
3              1.5           0.2
4              1.4           0.2
..             ...           ...
145            5.2           2.3
146            5.0           1.9
147            5.2           2.0
148            5.4           2.3
149            5.1           1.8

[150 rows x 2 columns]

#### Mengakses Kolom menggunakan Dot Notation
Penggunaan dot notation hanya bisa digunakan untuk mengakses 1 kolom saja sebagai Series. Dengan syarat kolom memiliki nama (tidak menggunakan indeks bilangan default). Nama kolom tidak boleh menggunakan nama atribut dari objek DataFrame. Nama kolom yang dapat diakses dengan dot juga tidak boleh mengandung karakter-karakter yang dilarang dalam nama variabel Python.

In [ ]:
# Mengakses kolom menggunakan dot notation
col_1 = df.PetalLengthCm
print(f"tipe objek: {type(col_1)}")
col_1

tipe objek: <class 'pandas.core.series.Series'>


0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: PetalLengthCm, Length: 150, dtype: float64

#### Mengakses Baris menggunakan Fungsi `loc` dan `iloc`
Kita dapat mengakses baris (dan kolom) DataFrame menggunakan loc atau iloc. `loc` mengakses indeks berdasarkan label dari baris atau kolom, sedangkan `iloc` mengakses indeks berdasarkan posisi dalam bilangan bulat. loc juga menerima range secara inklusif.

**Mengakses Baris**

In [ ]:
# Mengakses baris 1 sampai dengan 3 menggunakan loc
df.loc[1:3]  # range inklusif

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa

In [ ]:
# Mengakses baris 0 sampai dengan 3 menggunakan iloc
df.iloc[1:4] # range eksklusif

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa

**Mengakses Kolom**

In [ ]:
# Mengakses kolom sebagai Series dengan loc
df_1 = df.loc[:,'PetalLengthCm']
print("tipe objek:", type(df_1))
df_1

tipe objek: <class 'pandas.core.series.Series'>


0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: PetalLengthCm, Length: 150, dtype: float64

In [ ]:
# Mengakses kolom sebagai DataFrame dengan loc
df_2 = df.loc[:,['PetalLengthCm']]
print("tipe objek:", type(df_2))
df_2

tipe objek: <class 'pandas.core.frame.DataFrame'>


PetalLengthCm
0              1.4
1              1.4
2              1.3
3              1.5
4              1.4
..             ...
145            5.2
146            5.0
147            5.2
148            5.4
149            5.1

[150 rows x 1 columns]

In [ ]:
# Mengakses lebih dari 1 kolom dengan loc
df.loc[:,['PetalLengthCm', 'PetalWidthCm']]

PetalLengthCm  PetalWidthCm
0              1.4           0.2
1              1.4           0.2
2              1.3           0.2
3              1.5           0.2
4              1.4           0.2
..             ...           ...
145            5.2           2.3
146            5.0           1.9
147            5.2           2.0
148            5.4           2.3
149            5.1           1.8

[150 rows x 2 columns]

In [ ]:
# Mengakses kolom sebagai Series dengan iloc
df_3 = df.iloc[:,3]
print("tipe objek:", type(df_3))
df_3

tipe objek: <class 'pandas.core.series.Series'>


0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: PetalLengthCm, Length: 150, dtype: float64

In [ ]:
# Mengakses kolom sebagai DataFrame dengan iloc
df_4 = df.iloc[:,[3]]
print("tipe objek:", type(df_4))
df_4

tipe objek: <class 'pandas.core.frame.DataFrame'>


PetalLengthCm
0              1.4
1              1.4
2              1.3
3              1.5
4              1.4
..             ...
145            5.2
146            5.0
147            5.2
148            5.4
149            5.1

[150 rows x 1 columns]

In [ ]:
# Mengakses lebih dari 1 kolom dengan iloc
df.iloc[:,[2,4]]

SepalWidthCm  PetalWidthCm
0             3.5           0.2
1             3.0           0.2
2             3.2           0.2
3             3.1           0.2
4             3.6           0.2
..            ...           ...
145           3.0           2.3
146           2.5           1.9
147           3.0           2.0
148           3.4           2.3
149           3.0           1.8

[150 rows x 2 columns]

**Mengakses Baris dan Kolom**

In [ ]:
# Mengakses baris 5-10, Kolom PetalLengthCm dan PetalWidthCm menggunakan loc
df.loc[5:10, ['PetalLengthCm', 'PetalWidthCm']]

PetalLengthCm  PetalWidthCm
5             1.7           0.4
6             1.4           0.3
7             1.5           0.2
8             1.4           0.2
9             1.5           0.1
10            1.5           0.2

In [ ]:
# Mengakses baris 5-10, Kolom PetalLengthCm dan PetalWidthCm menggunakan iloc
df.iloc[5:11, [3,2]]

PetalLengthCm  SepalWidthCm
5             1.7           3.9
6             1.4           3.4
7             1.5           3.4
8             1.4           2.9
9             1.5           3.1
10            1.5           3.7

### A.2.8. Mengisi Missing Values
Nilai null atau missing values adalah informasi yang diisi dengan nilai null atau NaN (Not a Number). Missing values dapat disebabkan oleh berbagai hal, seperti kesalahan input, data yang tidak tersedia, atau data yang tidak relevan. Missing values dapat diisi dengan nilai lain, seperti mean, median, modus, atau nilai lain yang dianggap sesuai.

In [ ]:
df2 = pd.read_csv("disney_plus_titles.csv")

# menampilkan jumlah missing value per kolom
df2.isnull().sum()

show_id           0
type              0
title             0
director        473
cast            190
country         219
date_added        3
release_year      0
rating            3
duration          0
listed_in         0
description       0
dtype: int64

Dapat dilihat bahwa terdapat missing values pada kolom `director`, `cast`, `country`, `date_added`, dan `rating`. Jika ditampilkan cuplikan dataframe tersebut, maka akan terlihat bahwa missing values ditampilkan sebagai `NaN`.

In [ ]:
df2.head()

show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   

                                                cast        country  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...            NaN   
1           Jim Varney, Noelle Parker, Douglas Seale            NaN   
2  Raymond Albert Romano, John Leguizamo, Denis L...  United States   
3  Darren Criss, Adam Lambert, Derek Hough, Alexa...            NaN   
4  John Lennon, Paul McCartney, George Harrison, ...            NaN   

          date_added  release_year rating  duration  \
0  November 26, 2021          2016   TV-G    23 min   
1  November 26, 2021          1988     PG    91 min   
2  November 26, 2021          2011   TV-G    23 min   
3  November 26, 2021          2021  TV-PG    41 min   
4  November 25, 2021          2021    NaN  1 Season   

                       listed_in  \
0              Animation, Family   
1                         Comedy   
2      Animation, Comedy, Family   
3                        Musical   
4  Docuseries, Historical, Music   

                                         description  
0   Join Mickey and the gang as they duck the halls!  
1  Santa Claus passes his magic bag to a new St. ...  
2          Sid the Sloth is on Santa's naughty list.  
3               This is real life, not just fantasy!  
4  A three-part documentary from Peter Jackson ca...

Ada beberapa solusi yang dapat digunakan untuk mengatasi missing values, yaitu:
- Menghapus baris yang mengandung missing values, menggunakan fungsi `dropna()`
- Mengisi missing values dengan nilai tertentu, seperti mean, median, atau modus, menggunakan fungsi `fillna()`

#### Mengisi Missing Values menggunakan `fillna()` dengan nilai modus

In [ ]:
df2_fillna = df2.copy()

# mengisi missing value menggunakan modus setiap kolom
df2_fillna = df2_fillna.fillna(df2_fillna.mode().iloc[0])
df2_fillna.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

#### Menghapus Baris yang mengandung Missing Values menggunakan `dropna()`

In [ ]:
df2_drop_row = df2.copy()

# menghapus baris yang mengandung missing value
df2_drop_row = df2_drop_row.dropna()
df2_drop_row.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

## Pertanyaan Pandas
Jawaban boleh lebih dari 1 cell

### Nomor 1
Buka dataset `disney_plus_titles.csv` sebagai dataframe dan tampilkan 10 baris teratas dari dataframe tersebut.

In [2]:
# Jawab nomor 1

import pandas as pd

df_disney = pd.read_csv(
    "./disney_plus_titles.csv" # relative path dari file csv di lokal
)

df_disney.head(10)


show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   
5      s6    Movie                                 Becoming Cousteau   
6      s7  TV Show                                           Hawkeye   
7      s8  TV Show                            Port Protection Alaska   
8      s9  TV Show                         Secrets of the Zoo: Tampa   
9     s10    Movie             A Muppets Christmas: Letters To Santa   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   
5                         Liz Garbus   
6                                NaN   
7                                NaN   
8                                NaN   
9                   Kirk R. Thatcher   

                                                cast        country  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...            NaN   
1           Jim Varney, Noelle Parker, Douglas Seale            NaN   
2  Raymond Albert Romano, John Leguizamo, Denis L...  United States   
3  Darren Criss, Adam Lambert, Derek Hough, Alexa...            NaN   
4  John Lennon, Paul McCartney, George Harrison, ...            NaN   
5              Jacques Yves Cousteau, Vincent Cassel  United States   
6  Jeremy Renner, Hailee Steinfeld, Vera Farmiga,...            NaN   
7  Gary Muehlberger, Mary Miller, Curly Leach, Sa...  United States   
8  Dr. Ray Ball, Dr. Lauren Smith, Chris Massaro,...  United States   
9  Steve Whitmire, Dave Goelz, Bill Barretta, Eri...  United States   

          date_added  release_year rating   duration  \
0  November 26, 2021          2016   TV-G     23 min   
1  November 26, 2021          1988     PG     91 min   
2  November 26, 2021          2011   TV-G     23 min   
3  November 26, 2021          2021  TV-PG     41 min   
4  November 25, 2021          2021    NaN   1 Season   
5  November 24, 2021          2021  PG-13     94 min   
6  November 24, 2021          2021  TV-14   1 Season   
7  November 24, 2021          2015  TV-14  2 Seasons   
8  November 24, 2021          2019  TV-PG  2 Seasons   
9  November 19, 2021          2008      G     45 min   

                              listed_in  \
0                     Animation, Family   
1                                Comedy   
2             Animation, Comedy, Family   
3                               Musical   
4         Docuseries, Historical, Music   
5             Biographical, Documentary   
6           Action-Adventure, Superhero   
7         Docuseries, Reality, Survival   
8  Animals & Nature, Docuseries, Family   
9               Comedy, Family, Musical   

                                         description  
0   Join Mickey and the gang as they duck the halls!  
1  Santa Claus passes his magic bag to a new St. ...  
2          Sid the Sloth is on Santa's naughty list.  
3               This is real life, not just fantasy!  
4  A three-part documentary from Peter Jackson ca...  
5  An inside look at the legendary life of advent...  
6  Clint Barton/Hawkeye must team up with skilled...  
7  Residents of Port Protection must combat volat...  
8  A day in the life at ZooTampa is anything but ...  
9  Celebrate the holiday season with all your fav...

### Nomor 2
Tunjukkan jumlah nilai unik pada kolom `rating`. Kemudian sebutkan 3 jenis `rating` yang paling banyak muncul serta berapa banyak kemunculannya?

In [13]:
# Jawab nomor 2

df_disney.rating.value_counts().head(3)

TV-G     318
TV-PG    301
G        253
Name: rating, dtype: int64

### Nomor 3
Buatlah DataFrame baru yang terdiri atas kolom `title`, `release_year`, `rating`, dan `duration` untuk 30 baris pertama dari DataFrame pada **nomor 1**. Tampilkan dimensi/ukuran dan 3 baris pertama dari DataFrame baru tersebut.

In [14]:
# Jawab nomor 3

# Creates new df of 30 rows based on original df with new columns
df_new = df_disney[['title', 'release_year', 'rating', 'duration']].head(30).copy()

# Shows top 3
df_new.head(3)

title  release_year rating  \
0  Duck the Halls: A Mickey Mouse Christmas Special          2016   TV-G   
1                            Ernest Saves Christmas          1988     PG   
2                      Ice Age: A Mammoth Christmas          2011   TV-G   

  duration  
0   23 min  
1   91 min  
2   23 min

### Nomor 4
Tampilkan 5 movie dari **dataframe nomor 3** yang memiliki tahun rilis (release_year) paling awal (contoh 1990 lebih awal dibandingkan dengan 2000). *Hint: Silahkan eksplorasi terkait cara mengurutkan dataframe*

In [17]:
# Jawab nomor 4

# Sort by release_year
sorted_df = df_new.sort_values(by ="release_year", ascending= True)

# Shows top 5
sorted_df.head(5)

title  release_year rating duration
1                  Ernest Saves Christmas          1988     PG   91 min
19                              Enchanted          2007     PG  110 min
12                        The Pixar Story          2007      G   91 min
9   A Muppets Christmas: Letters To Santa          2008      G   45 min
2            Ice Age: A Mammoth Christmas          2011   TV-G   23 min

### Nomor 5
Tampilkan 1 movie dari **dataframe nomor 1** yang memiliki rating `PG` dan tahun rilis (release_year) paling awal.

In [5]:
# Jawab nomor 5

df_pg = df_disney[df_disney['rating'] == 'PG']
df_pg_awal = df_pg.sort_values(by='release_year').head(1)

df_pg_awal

show_id   type                  title       director  \
1227   s1228  Movie  Swiss Family Robinson  Edward Ludwig   

                                                   cast        country  \
1227  Thomas Mitchell, Edna Best, Freddie Bartholome...  United States   

             date_added  release_year rating duration  \
1227  November 12, 2019          1940     PG   94 min   

                               listed_in  \
1227  Action-Adventure, Family, Survival   

                                            description  
1227  A violent storm shipwrecks the Robinson family...

### Nomor 6
Tampilkan 5 movie pertama dari **dataframe nomor 1** yang tahun rilisnya (release_year) lebih dari rata-rata tahun rilis dari semua movie pada dataframe tersebut.

In [7]:
# Jawab nomor 6

df_mean = df_disney['release_year'].mean()
df_6 = df_disney[df_disney['release_year'] > df_mean]
df_6.head(5)


show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   
5      s6    Movie                                 Becoming Cousteau   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   
5                         Liz Garbus   

                                                cast        country  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...            NaN   
2  Raymond Albert Romano, John Leguizamo, Denis L...  United States   
3  Darren Criss, Adam Lambert, Derek Hough, Alexa...            NaN   
4  John Lennon, Paul McCartney, George Harrison, ...            NaN   
5              Jacques Yves Cousteau, Vincent Cassel  United States   

          date_added  release_year rating  duration  \
0  November 26, 2021          2016   TV-G    23 min   
2  November 26, 2021          2011   TV-G    23 min   
3  November 26, 2021          2021  TV-PG    41 min   
4  November 25, 2021          2021    NaN  1 Season   
5  November 24, 2021          2021  PG-13    94 min   

                       listed_in  \
0              Animation, Family   
2      Animation, Comedy, Family   
3                        Musical   
4  Docuseries, Historical, Music   
5      Biographical, Documentary   

                                         description  
0   Join Mickey and the gang as they duck the halls!  
2          Sid the Sloth is on Santa's naughty list.  
3               This is real life, not just fantasy!  
4  A three-part documentary from Peter Jackson ca...  
5  An inside look at the legendary life of advent...

### Nomor 7
Buatlah dataframe baru dari **dataframe nomor 1** dengan menghapus seluruh baris yang mengandung nilai null. Kemudian, tampilkan rata-rata tahun rilis (release_year) dari dataframe baru tersebut.

In [20]:
# Jawab nomor 7

# Delete all null values using .dropna()
df_7 = df_disney.dropna()

# Shows mean of Release year
df_7['release_year'].mean()

1999.8239608801955

### Nomor 8
Buatlah dataframe baru dari **dataframe nomor 1** dengan mengisi seluruh missing values menggunakan `modus`. Kemudian, tampilkan rata-rata tahun rilis (release_year) dari dataframe baru tersebut.

In [22]:
# Jawab nomor 8

# Get Mode
mode = df_disney.release_year.mode()

# Copy df Original
df_8 = df_disney.copy()

# Fill nulls with mode
df_8.release_year.fillna(mode, inplace= True)

# show new mean
df_8.release_year.mean()

2003.0917241379311

### Nomor 9
Apakah ada perbedaan antara rata-rata tahun rilis (release_year) dari **dataframe nomor 7** dan **dataframe nomor 8**? Jika ada, mengapa hal tersebut bisa terjadi? Jelaskan secara singkat.

> Jawaban:
Iya, terdapat perbedaan antara df_7 dan df_8, yaitu rata-rata antara kedua data frame berbeda. Hal ini disebabkan karena null values pada df_8 diganti dengan mode, sementara null values pada df_7 dihapus. Sehingga, jumlah nominal release_year akan berbeda alhasil nilai rata-rata yang berbeda.

# B. Pengenalan Numpy
[Numpy](https://numpy.org/) adalah library Python yang digunakan untuk komputasi numerik. Numpy menyediakan struktur data dan fungsi high-level untuk membuat pekerjaan dengan data numerik lebih mudah. Struktur data utama yang digunakan di Numpy adalah **array**. **Array adalah struktur data yang mirip dengan list, tetapi memiliki ukuran yang tetap dan hanya dapat berisi data dengan tipe yang sama**. Array pada Numpy dapat memiliki dimensi lebih dari 1, sehingga dapat digunakan untuk menyimpan data 2 dimensi atau lebih.

In [26]:
# Import library numpy
import numpy as np

## NumPy Array

### B.1.1. NumPy Array 1 Dimensi
**NumPy Array 1 Dimensi** adalah array 1 dimensi yang mirip dengan list. NumPy Array 1 Dimensi dapat dibuat dari list atau tuple.

In [27]:
# Membuat numpy array dengan list
arr = np.array([1,2,3,4,5])
arr

array([1, 2, 3, 4, 5])

In [28]:
print(type(arr))

<class 'numpy.ndarray'>


In [29]:
# Membuat numpy array dengan tuple
arr = np.array((1,2,3,4,5))
arr

array([1, 2, 3, 4, 5])

In [30]:
print(type(arr))

<class 'numpy.ndarray'>


Selain itu, terdapat juga beberapa fungsi yang dapat digunakan untuk membuat NumPy Array.

In [31]:
# Membuat array dengan nilai 0 sebanyak 4
print("np.zeros(4) : ",np.zeros(4))

# Membuat array dengan nilai 1 sebanyak 4
print("np.ones(4) : ",np.ones(4))

# Membuat array dengan nilai dalam range 1 sampai 15 dengan step sebesar 3
print("np.arange(1,15,3) : ",np.arange(1,15,3))

# Membuat array dengan nilai dalam range 1 sampai 10 sebanyak 4 dengan interval yang sama
print("np.linspace(1,10,4) : ",np.linspace(1,10,4))

np.zeros(4) :  [0. 0. 0. 0.]
np.ones(4) :  [1. 1. 1. 1.]
np.arange(1,15,3) :  [ 1  4  7 10 13]
np.linspace(1,10,4) :  [ 1.  4.  7. 10.]


### B.1.2. NumPy Array Multidimensi

Selain array 1 dimensi, NumPy juga mendukung array multidimensi.

In [32]:
# Membuat multidimensi array
multi_dim_arr = np.array([[1,2,3], [4,5,6], [7,8,9]])
multi_dim_arr

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

### B.1.3. Menampilkan Atribut Numpy Array

In [33]:
print(multi_dim_arr)

print()

# Menunjukkan ukuran array
print("multi_dim_arr.size:", multi_dim_arr.size)

# Menunjukkan jumlah dimensi atau rank dari sebuah array
print("multi_dim_arr.ndim:", multi_dim_arr.ndim)

# Mengetahui ukuran dari setiap dimensi array
print("multi_dim_arr.shape:", multi_dim_arr.shape)

[[1 2 3]
 [4 5 6]
 [7 8 9]]

multi_dim_arr.size: 9
multi_dim_arr.ndim: 2
multi_dim_arr.shape: (3, 3)


### B.1.4. Memeriksa Tipe Data
Kita dapat menggunakan fungsi `type()`.

In [34]:
# Mengecek tipe list
lst = [1,2,3,4,5]
print(f"type(lst): {type(lst)}")

# Mengecek tipe array
arr = np.array(lst)
print(f"type(arr): {type(arr)}")

type(lst): <class 'list'>
type(arr): <class 'numpy.ndarray'>


Selain fungsi `type()`, kita juga dapat menggunakan atribut `dtype` untuk mengetahui tipe data yang ada di dalam array.

In [35]:
# Mengecek tipe data dari array
print(f"arr.dtype: {arr.dtype}")

arr.dtype: int64


### B.1.5. Indexing, Slicing, dan Assigning Value pada NumPy Array

In [36]:
arr

array([1, 2, 3, 4, 5])

In [37]:
multi_dim_arr

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

#### Indexing

In [38]:
# Mengakses elemen indeks ke-3 dari array
arr[3]

4

In [39]:
# Mengakses elemen baris ke-1 dan kolom ke-2 dari multi dimensi array
print(multi_dim_arr[1,2])

# Cara lain
print(multi_dim_arr[1][2])

6
6


#### Slicing

In [40]:
# slice elemen dari indeks ke-2 sampai ke-4
arr[2:5]

array([3, 4, 5])

In [41]:
# Slice elemen dari baris kedua, kolom 1 sampai 3
multi_dim_arr[1,0:3]

array([4, 5, 6])

In [42]:
# Slice kolom kedua
multi_dim_arr[:,1]

array([2, 5, 8])

#### Assigning Value

In [43]:
# Mengubah value elemen ke-3 dari array
arr[2] = 10
arr

array([ 1,  2, 10,  4,  5])

In [44]:
# Mengubah value elemen baris ke-2 dan kolom ke-3 dari multi dimensi array
multi_dim_arr[1,2] = 15
multi_dim_arr

array([[ 1,  2,  3],
       [ 4,  5, 15],
       [ 7,  8,  9]])

### B.1.6. Operasi Matematika pada NumPy Array

In [45]:
arr_1 = np.array([1,2,3,4,5])
arr_2 = np.array([6,7,8,9,10])

In [46]:
# Menjumlahkan dua buah array
result = arr_1 + arr_2
result

array([ 7,  9, 11, 13, 15])

In [47]:
# Mengurangkan dua buah array
result = arr_1 - arr_2
result

array([-5, -5, -5, -5, -5])

In [48]:
# Menjumlahkan setiap elemen array dengan 10
result = arr_1 + 10
result

array([11, 12, 13, 14, 15])

In [49]:
# Mengalikan setiap elemen array dengan 10
result = arr_1 * 10
result

array([10, 20, 30, 40, 50])

In [50]:
# Perkalian dua buah array
result = arr_1 * arr_2
result

array([ 6, 14, 24, 36, 50])

In [51]:
# Dot product dua buah array
result = np.dot(arr_1, arr_2)
result

130

Operasi di atas dapat dilakukan pada array 1 dimensi maupun multidimensi. Kemudian, berikut ini adalah beberapa fungsi matematika yang dapat digunakan pada NumPy Array.

In [52]:
matrix = np.array([[1,2], [5,11]])
print(matrix)
print()

# Transpose matrix
print("Transpose:")
print(np.transpose(matrix))
print()

# Invers matrix
print("Invers:")
print(np.linalg.inv(matrix))

[[ 1  2]
 [ 5 11]]

Transpose:
[[ 1  5]
 [ 2 11]]

Invers:
[[11. -2.]
 [-5.  1.]]


### B.1.7. Operasi Statistik pada NumPy Array

In [53]:
arr = np.array([1,2,3,4,5,6,7,8,9,10,10,9,6,8,7,20,30,40,50,60,70,80,90,100])
arr

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  10,   9,   6,
         8,   7,  20,  30,  40,  50,  60,  70,  80,  90, 100])

In [54]:
# Nilai rata-rata
print("Rata-rata:", np.mean(arr))

# Nilai median
print("Median:", np.median(arr))

# Nilai maksimum
print("Maksimum:", np.max(arr))

# Nilai standar deviasi
print("Standar deviasi:", np.std(arr))

# Nilai percentile
print("Percentile 75:", np.percentile(arr, 75))

# Nilai unik
print("Nilai unik:", np.unique(arr))

Rata-rata: 26.458333333333332
Median: 9.0
Maksimum: 100
Standar deviasi: 30.49177370847568
Percentile 75: 42.5
Nilai unik: [  1   2   3   4   5   6   7   8   9  10  20  30  40  50  60  70  80  90
 100]


## Pertanyaan Numpy
Jawaban boleh lebih dari 1 cell

### Nomor 1

In [55]:
arr_1 = np.array([1, 0, np.nan, 1, 100, 90, 1000, 'a', 'b', 'c', 'hello world'])
arr_1.dtype

dtype('<U32')

Jalankan cell di atas! Mengapa `arr_1.dtype` bertipe `<32` padahal terdapat elemen integer dan NaN?

> Jawaban:
Mengingat bahwa array hanya boleh mempunyai 1 tipe data saja, NumPy mencoba menemukan tipe data yang konsisten antara semua elemen dalam array dan jika tidak dapat menemukan tipe data numerik tunggal yang konsisten, maka tipe data yang dipilih akan menjadi tipe data string Unicode dengan panjang tetap, seperti 'U32'.

### Nomor 2
Diketahui beberapa array multidimensi sebagai berikut.

In [56]:
A = np.array([[1,2,0],[5,2,1],[3,4,5]])
B = np.array([[5,0,2],[0,3,1],[7,0,5]])
C = np.array([[1,1,2],[2,3,1],[2,2,2]])

 Dari ketiga multidimensi array di atas, lakukanlah langkah-langkah berikut ini.
1. Penjumlahan antara A dan C
2. Buat matrix baru yang isinya adalah hasil akar (square root) tiap elemen dari matrix pada langkah 1
3. Jalankan operasi invers pada matrix pada langkah 2
4. Terapkan operasi dot product pada matrix pada langkah 3 dengan matrix B
5. Jalankan perkalian skalar pada matrix A dengan hasil dot product pada langkah 4
*hint*: anda dapat menggunakan fungsi `sqrt()`

Secara matematis, langkah di atas dapat dituliskan sebagai berikut.
1. $V = A + C$
2. $W =$ sebuah matrix yang tiap elemennya adalah elemen V diakarkan
3. $X = W^{-1}$
4. $K = X\cdot{B}$
5. $result = K*A$

In [71]:
# Jawab nomor 2

# 1. Penjumlahan antara A dan C
V = A + C
print("1. A + C:")
print(V)
print()

# 2. 
W = np.sqrt(V)
print("2. sqrt(V):")
print(W)
print()

# 3. 
X = np.linalg.inv(W)
print("3. W-1:")
print(X)
print()

# 4.
K = np.dot(X, B)
print("4. X . B:")
print(K)
print()

# 5.
result = A * K
print("5. KA:")
print(result)





1. A + C:
[[2 3 2]
 [7 5 2]
 [5 6 7]]

2. sqrt(V):
[[1.41421356 1.73205081 1.41421356]
 [2.64575131 2.23606798 1.41421356]
 [2.23606798 2.44948974 2.64575131]]

3. W-1
[[-2.25900135  1.03044737  0.65668972]
 [ 3.53568399 -0.53378109 -1.60458459]
 [-1.3642027  -0.37670172  1.30851703]]

4. X . B:
[[-6.69817868  3.09134211 -0.20410671]
 [ 6.4463278  -1.60134328 -1.48533608]
 [ 2.33860568 -1.13010515  3.43747801]]

5. KA:
[[-6.69817868  6.18268423 -0.        ]
 [32.23163899 -3.20268655 -1.48533608]
 [ 7.01581703 -4.52042059 17.18739006]]


### Nomor 3

In [72]:
arr_3 = np.array([1,2,3,4,5,6,7,8,9,10,5,2,4,1,58,2,10,38,10,11,12,13,14,15,16,17,18,19,20])

Kalikan `arr_3` dengan 11, kemudian hitung dan tampilkan rata-rata, median, nilai maksimum, standar deviasi, variance, dan percentile 10!

In [73]:
# Jawab nomor 3

new_arr = arr_3 * 11

print("Rata-rata: ", np.mean(new_arr))
print("Median: ", np.median(new_arr))
print("Maksimum: ", np.max(new_arr))
print("Standar deviasi: ", np.std(new_arr))
print("Variance: ", np.var(new_arr))
print("Percentile 10: ", np.percentile(new_arr,10))

Rata-rata:  128.9655172413793
Median:  110.0
Maksimum:  638
Standar deviasi:  128.24504635418725
Variance:  16446.791914387635
Percentile 10:  22.0


#